include/picongpu/plugins/common/openPMDAttributes.hpp

```diff
@@ -0,0 +1,140 @@
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+
+#include <openPMD/openPMD.hpp>
+
+#include <string>
+#include <vector>
+
+namespace picongpu
+{
+    namespace openPMD
+    {
+        class SetMeshAttributes
+        {
+        private:
+            static std::vector<float_64> initGridGlobalOffset(uint32_t currentStep);
+            static std::vector<float_X> initGridSpacing();
+            static std::vector<std::string> initAxisLabels();
+
+        public:
+            static constexpr ::openPMD::UnitDimension openPMDUnitDimensions[7]
+                = {::openPMD::UnitDimension::L,
+                   ::openPMD::UnitDimension::M,
+                   ::openPMD::UnitDimension::T,
+                   ::openPMD::UnitDimension::I,
+                   ::openPMD::UnitDimension::theta,
+                   ::openPMD::UnitDimension::N,
+                   ::openPMD::UnitDimension::J};
+
+            // Attributes per Mesh
+            std::vector<std::string> m_axisLabels = initAxisLabels();
+            ::openPMD::Mesh::DataOrder m_dataOrder = ::openPMD::Mesh::DataOrder::C;
+            ::openPMD::Mesh::Geometry m_geometry = ::openPMD::Mesh::Geometry::cartesian;
+            std::vector<float_64> m_gridGlobalOffset; // no default since it depends on time step
+            std::vector<float_X> m_gridSpacing = initGridSpacing();
+            double m_gridUnitSI = UNIT_LENGTH;
+            float_X m_timeOffset = 0.;
+            std::map<::openPMD::UnitDimension, double> m_unitDimension{};
+
+            // Attributes per Mesh Component
+            // my compiler thinks I want to define a function when I use the shorthand
+            std::vector<float_X> m_position = std::vector<float_X>(simDim, 0.5);
+            double m_unitSI = 1.0;
+
+            SetMeshAttributes(uint32_t currentStep);
+
+            SetMeshAttributes const& operator()(::openPMD::Mesh& mesh) const;
+
+            SetMeshAttributes const& operator()(::openPMD::MeshRecordComponent& dataset) const;
+        };
+
+
+        /******************
+         * IMPLEMENTATION *
+         ******************/
+
+        std::vector<std::string> SetMeshAttributes::initAxisLabels()
+        {
+            using vec_t = std::vector<std::string>;
+            return simDim == DIM2 ? vec_t{"y", "x"} : vec_t{"z", "y", "x"};
```

Comment:
Note you can simplify the code and remove the declaration above `return simDim == DIM2 ? {"y", "x"} : {"z", "y", "x"};`

Reply:
That's what I had tried first, but apparently C++ cannot do type inference in this statement:
```
/home/franzpoeschel/git-repos/picongpu/include/pmacc/../picongpu/plugins/common/openPMDAttributes.hpp(79): error #29: expected an expression

/home/franzpoeschel/git-repos/picongpu/include/pmacc/../picongpu/plugins/common/openPMDAttributes.hpp(79): error #53: expected a ":"
```

Reply:
It cannot because initializer lists are only for objects. So you could have (but of course should not have) expressed it as
```cpp
return simDim == DIM2 ? std::initializer_list<std::string>({"y", "x"}) : std::initializer_list<std::string>({"z", "y", "x"});
```